# <FONT COLOR="blue"> **Samsung Innovation Campus**</FONT>  

## <FONT COLOR="blue">**Codificación y Programación**</FONT>

# **Capítulo 3: Programación eficaz en Python: Función, Clausura y Clase**

## Objetivos del capítulo

Los alumnos podrán utilizar funciones y técnicas de programación modular. Los alumnos podrán usar expresiones lambda, que son expresiones que se usan para expresar funciones simples. Los alumnos podrán comprender el cierre que recibe el valor de retorno como una función después de definir funciones anidadas. Además, los alumnos podrán definir clases, que son conceptos importantes de los lenguajes orientados a objetos.

## Contenido del capítulo
- Unidad 17. Función
- Unidad 18. Llamado Recursivo de una Función
- Unidad 19. Lambda
- Unidad 20. Clausura
- Unidad 21. Clase

# **Unidad 18. Funciones recursivas**

## Objetivos de aprendizaje
- Entender y ser capaz de usar funciones recursivas.
- Entender el concepto de factorial y ser capaz de implementarlo a través de declaraciones iterativas.
- Ser capaz de implementar las funciones factorial y Finobacci a través de funciones recursivas en vez de usar loops.
- Entender las deficiencias de las funciones recursivas y ser capaz de implementar la secuencia de Finobacci más rápidamente a través de la memorización.

## Resumen de la lección
- Aprenderá cómo resolver problemas a través de llamar a la función así misma.
- Aprenderá a cómo implementar la función factorial a través de loops para verificar la eficiencia de las funciones recursivas.
- Aprenderá como crear una función recursiva y llamarla para encontrar el factorial y los números de la secuencia de Finobacci.
- Aprenderá las desventajas y limitaciones de las llamadas recursivas y cómo implementar la secuencia de Finobacci de mejor forma a través de la memorización.
 
## Conceptos previos de debes saber de las unidades previas
- Saber como definir y llamar una función definida por el usuario usando la declaración *def*.
- Saber como ingresar valores a las funciones eficientemente a través del uso de los parámetros variables, parámetros por defecto y los parámetros por palabras clave.
- Saber como retornar múltiples valores a través de la declaración *return* de una función.


## Palabras clave

- Llamada recursiva.
- Factorial.
- Secuencia de Finobacci.
- Memorización.

## <FONT COLOR="blue"> Misión </FONT>

### <FONT COLOR="blue">1. Problema del mundo real </FONT>


#### **1.1 ¿Cómo hacemos un fractal?**

<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig1.png" width="880px" />

Los humanos han usado computadores para resolver problemas complejos. Sin embargo, aún hay problemas difíciles en este mundo. En la figura de a continuación podemos observar un ejemplo representativo de lo que es un fractal. Un fractal es un fenómeno natural que exhibe características que se repiten con cierta regularidad y puede ser usado, por ejemplo, para crear imágenes de la naturaleza tales como montañas y nubes en efectos especiales de los videos. De hecho, algunos efectos especiales en gráficas puede ser creados haciendo uso de los fractales. En programación, es complejo expresar una figura fractal para que lo entienda un computador. Sin embargo, hay un camino para resolver este problema a través de la **Recursividad**.

#### **1.1 El problema de las torres de Hanoi**
<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig2.png" width="480px" height="240px" />

El problema de la torre de Hanoi es un problema representativo del uso de las funciones recursivas para solucionar de forma sencilla problemas complejos. La torre de Hanoi es un juego en el cual los discos mostrados en la figura son movidos a los bastones de acuerdo a unas reglas. Cuando se está jugando, existe un número óptimo de movimientos para su solución que varía según la cantidad de discos que hayan en el juego. Si hay n discos, el número óptimo de movimientos es 2^n - 1. Vamos a aprender de estos conceptos mientras realizamos un programa para este problema.

### <FONT COLOR="blue"> 2. Solución </FONT>
#### **2.1 Solucionando el rompecabezas de la torre de Hanoi**
<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig3.png" width="480px" height="240px" />

La torre de Hanoi también es un juego de rompecabezas. Hay 3 bastones y 3 discos de diferente tamaño que encajan en estos bastones. Antes de empezar el rompecabezas, los discos son agrupados en un bastón de manera ordenada dejando el más pequeño en la parte superior. El objetivo del juego es mover los discos de un bastón a otro de tal forma que se satisfagan las condiciones:
- Mover los n-discos de diferente tamaño desde un punto A a un punto final C.
- Solo un disco puede ser movido a la vez.
- Cuando se mueve un disco, lo pasas de la cima de un bastón y solo lo puedes mover a la cima del otro bastón.
- En el proceso de mover los discos, un disco grande no puede ser puesto por encima de un disco más pequeño.
Para profundizar puedes visitar la página: [clic aqui](https://en.wikipedia.org/wiki/Tower_of_Hanoi)

### <FONT COLOR="blue"> 3. Misión </FONT>

#### **3.1 Algoritmo de las torres de Hanoi**
<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig4.png" width="480px" height="240px" />

Algoritmo:
- Si hay un solo disco, este se puede mover desde un punto a otro en un turno.
- Asuma que hay 3 discos y 3 bastones:
    - Con el fin de mover el disco menor las bastón destino, mueva 3-1 discos a los bastones auxiliares.
    - Mueva el disco 3 al bastón inicial.
    - Mueva los 3-1 discos de los bastones auxiliares al bastón de destino.
Podemos resolver este problema a través del método recursivo.

#### **3.2 y 3.3 Función Hanoi y Cómo funciona el algoritmo para el caso de que n=3**
En la función, en la parte del *else* resuelve el problema llamándose a a si misma la función hanoi() para mover el disco. Esta es la parte central del código. Ahora vamos a programar esta función:

In [1]:
def hanoi(n, from_, to_, via_):
    if n == 1:
        print('Numero {}, {} -> {}'.format(n, from_, to_))
    else:
        hanoi(n-1, from_, via_, to_)
        print('Numero {}, {} -> {}'.format(n, from_, to_))
        hanoi(n-1, via_,to_, from_)
        
hanoi(3, 'A', 'C', 'B')

Numero 1, A -> C
Numero 2, A -> B
Numero 1, C -> B
Numero 3, A -> C
Numero 1, B -> A
Numero 2, B -> C
Numero 1, A -> C


#### **3.4 Pseudocódigo**
A continuación veremos el pseudocódigo y su correspondiente diagrama de flujo. En el cual, a grandes rasgos, empieza llamando la función hanoi(n, from_, to_, via_), si solo queda un disco entonces mueve el disco de from_ a to_. En caso de que no, muevel el disco n-1 desde from_ a to_ a través de la función de hanoi(n-1, from_, via_, to_). Luego, mueve el disco n (el más grande) from_ to_ el destino. Después, mueve el disco n-1 en via_ hacía el destino, from_ es usado como un bastón auxiliar. Después, llame a la función hanoi(n-1, via_, to_, from_) y finaliza.

<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig5.png" width="880px" />

## <FONT COLOR="blue"> Concepto Clave </FONT>
### <FONT COLOR="blue"> 1. Encuentre el factorial usando ciclos </FONT>
#### **1.1 Definición de factorial**

El factorial es un término matemático y se expresa como $n!$ para un número natural $n$. El factorial es el producto de todos los números naturales desde 1 hasta $n$ el cual es un número natural cualquiera, matemáticamente su fórmula es:

<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig6.png" width="880px" />

#### **1.2 Factorial usando el ciclo for**
Para realizarlo vamos a programar el factorial usando el ciclo for como se muestra a continuación.

In [2]:
#n = int(input('Enter a number: '))
n=5
fact = 1
for i in range(1, n+1):
    fact = fact*i
    
print('{}! = {}'.format(n, fact))

5! = 120


### <FONT COLOR="blue"> 2. Encuentre el factorial usando una función recursiva </FONT>
#### **2.1 ¿Qué es una función recursiva?**

Una llamada recursiva es una función que se llama asi misma dentro de la función. La recursividad es una técnica de solución de problemas muy útil, puede resolver problemas de manera simple e intuitiva que serían más complejos si se solucionan con técnicas procedimentales. En este caso en particular,

$$n! = n*(n-1)!$$ y si 
$$n<=1 \rightarrow 1$$

Revisemos el siguiente ejemplo:

<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig7.png" width="880px"  />

#### **2.2 El flujo de una función recursiva**
El flujo de la función factorial recursiva se muestra a continuación. Factorial(1) regresa 1. luego, por factorial(2), el cual retorna: *2*factorial(1) = 2* por 1 y asi sucesivamente, hasta que finalmente retorna 120.

<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig8.png" width="880px"  />

Hay que tener en cuenta que en una función recursiva, la condición de terminación es completamente necesaria. Por otro lado, por definición el factorial de 1! = 1, por lo que no habría necesidad de definir un valor de factorial para valores menores a este. Sin embargo, en este caso se define y cuando el valor ingresado es menor a 1, el programa regresa 1. En código, esto se ve de la siguiente manera:

In [3]:
def factorial(n): # Función recursiva para n!
    if n <= 1:    # Condición de terminación 
        return 1
    else:
        return n*factorial(n-1) # n*(n-1)! implementación de la definición
    
n = 5
print('{}! = {}'.format(n, factorial(n)))

5! = 120


**Un paso adicional, la recursividad y el algoritmo de divide y vencerás**

En la recursividad, cuando una función se llama así misma y el tamaño del problema es n, es una práctica común es dividirla en pequeños problemas menores que n. Este método de solución está cercanamente relacionado con el algoritmo de divide y conquista. Este algoritmo, puede ser llamado como una recursión multi-rama. El ejemplo del factorial mostrado anteriormente no es un algoritmo de divide y conquista porque es una recursión de una rama donde la llamada recursiva solo se realiza una vez. Cuando el tamaño del problema es n y el problema se divide en dos de tamaño n/2 y se pasan los argumentos recursivamente, a esto, si se consideraría como un caso típico del algoritmo de divide y conquista. Adicionalmente a los casos en los que se reparten en dos, los métodos que se reparten en más de dos problemas más pequeños, son considerados como algoritmos de divide y conquista (en un problema de divide y conquista, hay una **restricción**: La suma de los problemas pequeños no debe ser más grande que el problema original). Finalmente, si aprendes estructuras de datos después, aprenderás un algoritmo llamado ordenamiento rápido (quick sort). Este algoritmo puede ser implementado con el método de divide y conquista a través de llamados recursivos.

### <FONT COLOR="blue"> 3. Desventajas de las funciones recursivas y de memorización </FONT>
#### **3.1 Ventajas y desventajas de las funciones con ciclos y recursivas**

A continuación, los pros y contras de las funciones recursivas

<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig9.png" width="880px" />

#### **3.2 Definición de la serie de Finobacci**

Una serie de Finobacci es una secuencia en la que el primer el segundo término son 1 y los términos subsiguientes son la suma de los términos inmediatamente anterior. 

El primer término $F_0$ de la serie es 0, $F_1$ es 1 y $F_n$ es definido como sigue:

$$F_n = F_{n-1} + F_{n-2}$$

De acuerdo a su definición, la serie de Finobacci es:

$$0, 1, 1, 2, 3, 5, 8, 13, 21, ... $$

La serie de Finobacci puede crear secuencias con patrones geométricos como se muestra en la siguiente figura.

<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig10.png" width="880px" />

A continuación se muestra la función recursiva de Finobacci

In [4]:
def finobacci(n): # Función recursiva
    if n <= 1:
        return n
    else:
        return (finobacci(n-1) + finobacci(n-2)) # F_n = F_(n-1) + F_(n-2)
    
#nterms = int(input("Cuántos términos de la secuencia de Finobacci quieres: "))
nterms = 5
# No se puede generar los términos si n es negativo
if nterms <= 0:
    print('Error, ingrese un número positivo.')
else:
    print('La secuencia de Finobacci: ', end = ' ')
    for i in range(nterms):
        print(finobacci(i), end = ' ')
        

La secuencia de Finobacci:  0 1 1 2 3 

#### **3.3 Proceso de ejecución de la función de Finobacci**

En este literal vamos a profundizar lo que pasa internamente de la función finobacci(5). Como se observa en la imagen, fib(5) llama a fib(4) y fib(3), y fib(4) llama recursivamente a fib(3) y fib(2). Este proceso resulta en una ejecución muy larga en forma de árbol.
<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig11.png" width="480px" height="240px" />

#### **3.4 Implementación de la función de Finobacci**
La implementación de la serie de Finobacci de manera recursiva se define a continuación, en la que el valor se adiciona recursivamente hasta que se cumpla el criterio de terminación.

In [5]:
def finobacci(n): # Función recursiva
    if n==0:
        return 0
    elif n == 1:
        return 1
    else:
        return (finobacci(n-1) + finobacci(n-2)) # F_n = F_(n-1) + F_(n-2)

print(finobacci(10))

55


#### **3.5 Eficiencia de la implementación de la función de Finobacci como función recursiva**

El mayor problema con las funciones recursivas es que no es eficiente porque repite cálculos sin importar que ese valor incluso ya se ha calculado previamente. Miremos cuanto tiempo le toma al programa encontrar el término 30 de la secuencia de finobacci, para ello vamos a usar el módulo para la medición de tiempo *Timer()* que provee esta métrica. Para profundizar puede consultar le módulo [timeit](https://docs.python.org/3/library/timeit.html).

In [6]:
from timeit import *
def finobacci_1(n): # Función recursiva
    if n==0:
        return 0
    elif n == 1:
        return 1
    else:
        return (finobacci_1(n-1) + finobacci_1(n-2)) # F_n = F_(n-1) + F_(n-2)

t3 = Timer("finobacci_1(30)", "from __main__ import finobacci_1")
print("finobacci_1(30)*20 times: ", t3.timeit(number = 20), "segundos")

finobacci_1(30)*20 times:  6.822911200000001 segundos


Como vimos en el ejemplo anterior, la tarea se repite 20 veces y toma aproximadamente 9.7 segundos. Si escribimos números por encima de estos probablemente tome mucho más tiempo.

#### **3.6 Función de Finobacci y memoización**

La memorización es un método de almacenamiento de los resultados de los pequeños problemas que se repiten una y otra vez pero que su resultado no cambia. Para resolver los problemas que hemos visto a lo largo de la lección lo más recomendable sería usar, *memoización* y *Programación dinámica (DP)*. La programación dinámica es una de los algoritmos de solución de problemas que veremos en el curso de algoritmos posteriormente. Sin embargo, algunos puntos importantes de este método:
- En programación dinámica tienes que resolver pequeños problemas para resolver los problemas grandes.
- Los resultados de estos pequeños problemas son usualmente constantes.
- Los problemas pequeños que no cambian ocurren repetidamente.

En un problema específico, en el que los resultados son almacenados a través de la memorización, los cálculos innecesarios pueden ser omitidos y devolver, de manera más rápida, los valores almacenados. En una función recursiva se tiene que también resolver pequeños problemas hasta el criterio de finalización es cumplido y lo más probablemente es que estos pequeños problemas se repitan a lo largo del proceso.

A continuación se muestra un ejemplo de introducción a la memorización para la función de fib(5) para encontrar el término 5 de la serie de Finobacci. En la imagen de la izquierda, es el caso donde se realiza el llamado a 15 funciones ya que no se hace uso de la memorización. Por otro lado, en la imagen de la derecha, los llamados recursivos se reducen a 9 debido a que algunos de los llamados de la función se habían calculado previamente y se habían almacenado. El proceso de memorización se puede realizar a través de un diccionario de python.

<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig12.png" width="880px" />

#### **3.7 Función de Finobacci y el principio de memoización**
Con el fin de evitar el doble cálculo, si los valores de los términos son almacenados en el diccionario *dic*, entonces los valores una vez calculados no se vuelven a re-calcular. A continuación se muestra la función que implementa esta estrategia.

In [7]:
from timeit import *

dic = {0:0, 1:1}
def finobacci_2(n): # Función recursiva
    if n in dic:
        return dic[n]
    
    dic[n] = finobacci_2(n-1) + finobacci_2(n-2)
    return dic[n]

t2 = Timer("finobacci_2(30)", "from __main__ import finobacci_2")
print("finobacci_2(30)*20 times: ", t2.timeit(number = 20), "segundos")

finobacci_2(30)*20 times:  2.590000000068926e-05 segundos


Si recordamos, en el caso que no teníamos implementado la memorización tomó 9.7 segundos, pero con la memorización, este resultado se optimizó a 0.000033 segundos. Por lo tanto, si se reduce la eficiencia de una función recursiva, se debe considerar incrementar su eficiencia a través de almacenar los resultados cada vez que la función recursiva se llama. Finalmente, se muestra el diagrama de flujo de como los diccionarios son almacenados.

<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig13.png" width="880px" />


### <FONT COLOR="blue"> 4. Repaso de la lección </FONT>
### Q1. *(tiempo estimado: 5 min)*
Toma ue número n como entrada y encuentre su suma desde 1 hasta n. Escriba su función usando funciones recursivas.
### Q2. *(tiempo estimado: 5 min)*
Python tiene el operador ** el cual indica una potencia, Sin embargo, tomemos a x y n como entradas sin usar ese operador y use una función recursiva que retorne x a la potencia de n. Luego, calcule $2^{10}$ poniendo x como 2 y n como 10.

## <FONT COLOR="blue"> Vamos a codificar </FONT>
### <FONT COLOR="blue"> 1. Funciones pre-construidas </FONT>
#### **1.1 Funciones pre-construidas expresadas como cajas negras**

Hasta el momento hemos estado usando funciones como *print* e *input*. Ahora, aprenderemos mas acerca de estas. En programación, una función es a menudo tratada como una caja negra. Esto es porque el usuario que la llama solo necesita proveerle de la entrada correcta y este solo requiere su valor resultante.

<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig14.png" width="480px"  />

Por ejemplo, desde que sabemos que la función *len*  permite contar el número de caracteres de una cadena de texto, nosotros la usamos sin conocer el detalle interno de ese proceso. De manera similar, las funciones *print* e *input* pertenecen a este grupo de funciones. Las funciones que son implementadas por defecto en Python son llamadas como funciones pre-construidas y Python proveen más de 100 de estas.

<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig15.png" width="480px"  />

#### **1.2 Funciones pre-construidas de Python**
<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig16.png" width="880px" />

#### **1.3 Cómo usar funciones pre-construidas de Python**

Las funciones que usan código pre-construido por Python no requieren ser importadas. Por ejemplo:

In [8]:
# La función abs retorna el valor absoluto
abs(-100)

100

In [9]:
# La función min retorna el valor mínimo de los elementos ingresados
min(200, 100, 300, 400)

100

In [10]:
# La función max retorna el valor maximo de los elementos ingresados
max(200, 100, 300, 400)

400

#### **1.4 El valor retornado de una función pre-construida de Python**

In [11]:
str1 = 'F00'
len(str1) # Retorna la longitud de un string

3

In [12]:
type(str1) # Retorna el tipo de un objeto 

str

In [13]:
id(str1) # Retorna el ID del valor de un objeto

2490601741040

In [14]:
eval("100+200+300") # Convierte la cadena de texto en valores y operaciones numéricas y luego las evalua

600

#### **1.5 Funciones pre-construidas de Python asociadas al ordenamiento**

El código ordena un listado de cadenas de texto y enteros de forma ascendente y descendente

In [15]:
sorted("EABFD") # Ordenamiento de una cadena de texto

['A', 'B', 'D', 'E', 'F']

In [16]:
n_list = [200, 100, 300, 400, 50]
sorted(n_list) # Ordenamiento de un listado en orden ascendente

[50, 100, 200, 300, 400]

In [17]:
sorted(n_list, reverse=True) # Ordenamiento de un listado en orden descendente

[400, 300, 200, 100, 50]

#### **1.6 La función *id* y la programación orientada a objetos**

Python es un lenguaje de programación orientada a objetos. El paradigma en el que objetos con diversas propiedades y funciones componen un programa es el núcleo de la programación orientada a objetos. Los objetos de Python tienen una única identidad que las distingue de otros objetos. La función id retorna la identidad de un objeto como un número entero. Los objetos se tratarán en la unidad 21. Por el momento, tenemos los siguientes ejemplos:  

In [18]:
a_str = 'Hello Python!'
id(a_str)

2490601742064

In [19]:
n = 300
id(n)

2490618589968

#### **1.7 La función *type***
La función *type* retorna el tipo de dato de un objeto. Practiquemos un poco este concepto.

In [20]:
type(123)

int

In [21]:
type('Hello string!')

str

In [22]:
type(120.3)

float

In [23]:
type([100, 300, 600])

list

#### **1.8 La función *eval***
La función *eval* es una función donde el intérprete de Python ejecuta el valor de una cadena de texto como argumento y cuya salida es el resultado que se espera de los cálculos descritos en ella. Esta función es muy poderosa, como la ejecución del script en sí. Pero debe ser usada con cuidado porque puede llamar directamente comandos del sistema para obtener información de todas las carpetas en el sistema donde Python se está ejecutando e incluso puede borrarlos. Veamos algunos ejemplos:

In [24]:
eval('10+20') # El traductor de Python ejecuta la sentencia 10 + 20

30

In [25]:
eval('(5*20)/2') # El traductor de Python ejecuta la sentencia (5*20)/2

50.0

In [26]:
chr(65) # Retorna el valor Unicode de 65 el cual es la letra del alfabeto 'A'

'A'

In [27]:
ord('A') # Retorna el valor Unicode de la letra del alfabeto 'A'

65

### <FONT COLOR="blue"> 2. El problema de la torre de Hanoi </FONT>
#### **2.1 Proceso de ejecución paso por paso**

Ahora vamos a profundizar en el proceso de ejecución del algoritmo de la torre de Hanoi. Para resolverlo usaremos funciones recursivas.

- Considere el caso con 3 discos con se muestra en la figura. En la etapa final, el disco más grande (el rojo) debe ir de A a C.
- Para ello, los discos del medio (el azul y el verde) deben ir a B.
- Se debe considerar mover los dos discos al destino B usando como punto de comienzo en A y el bastón C sin considerar el disco grande.
- Cuando todos los discos más pequeños se han movido a B, el disco más grande (el rojo) se mueve a C [Figura 3].
- En este caso, habrá dos discos pequeños en el bastón B [Figura 4].
- Después de eso, se debe considerar mover los dos discos más pequeños desde el punto B al punto C usando el bastón de apoyo A [Figuras 5, 6 y 7].

<img src="https://raw.githubusercontent.com/EdgarGM20/Imagenes/main/Ch18_fig4.png" width="680px"  />

Cuando hay solo 1 disco, este se puede mover directamente de A a C.

Cuando hay 2 discos,
- Primero se mueve el disco pequeño a B (Como si fuera solo 1 disco)
- Luego, se mueve el más grande a C.
- Finalmente, se mueve el el disco pequeño de B a C. (Como si fuera solo 1 disco)

Cuando hay 3 discos,
- Primero mueva el (disco 1 y 2) de A a B. (Como si fueran solo 2 discos)
- Luego, mueva el disco 3 a C cuando solo quede un disco en esta bastón.
- Finalmente mueva el (disco 1 y 2) de B a C. (Como si fueran solo 2 discos)

Cuando hay n-discos,
- Primero mueva los n-1 discos a B. (Como si fueran n-1 discos)
- Luego mueva el último disco a C cuando solo quede un disco en este bastón.
- Finalmente, mueva los n-1 discos de B a C (Como si fueran n-1 discos)

Por lo que finalmente, se tendría el siguiente código:

In [28]:
def hanoi(n, from_, to_, via_):
    if n == 1:
        print('Numero {}, {} -> {}'.format(n, from_, to_))
    else:
        hanoi(n-1, from_, via_, to_)
        print('Numero {}, {} -> {}'.format(n, from_, to_))
        hanoi(n-1, via_,to_, from_)
        
hanoi(3, 'A', 'C', 'B')

Numero 1, A -> C
Numero 2, A -> B
Numero 1, C -> B
Numero 3, A -> C
Numero 1, B -> A
Numero 2, B -> C
Numero 1, A -> C


## <FONT COLOR="blue">Programación por parejas </FONT>

### <FONT COLOR="blue">Práctica de la programación en parejas </FONT>


**Directriz, mecanismos y plan de contingencia**

 Preparar la programación por parejas implica establecer pautas y mecanismos para ayudar a los alumnos a emparejarse correctamente y mantenerlos emparejados.  Por ejemplo, los estudiantes deben turnarse para "conducir el mouse".  La preparación efectiva requiere planes de contingencia en caso de que uno de los socios se ausente o decida no participar por una u otra razón.  En estos casos, es importante dejar claro que el alumno activo no será sancionado porque el emparejamiento no funcionó bien.
 
**Emparejar habilidades similares, no necesariamente iguales, como socios**

 La programación en parejas puede ser efectiva cuando los estudiantes con habilidades similares, aunque no necesariamente iguales, se emparejan como compañeros.  El emparejamiento de estudiantes que no coinciden a menudo puede conducir a una participación desequilibrada.  Los maestros deben enfatizar que la programación en parejas no es una estrategia de "divide y vencerás", sino más bien un verdadero esfuerzo de colaboración en cada esfuerzo por todo el proyecto.  Los maestros deben evitar emparejar a estudiantes muy débiles con estudiantes muy fuertes.

**Motivar a los estudiantes ofreciendo incentivos adicionales**

 Ofrecer incentivos adicionales puede ayudar a motivar a los estudiantes a emparejarse, especialmente con estudiantes avanzados.  A algunos maestros les ha resultado útil pedirles a los estudiantes que se emparejen solo para una o dos tareas.

**Evitar trampas en la colaboración**

 El desafío para el maestro es encontrar formas de evaluar los resultados individuales, mientras aprovecha los beneficios de la colaboración.  ¿Cómo saber si un estudiante aprendió o hizo trampa?  Los expertos recomiendan revisar el diseño y la evaluación del curso, así como discutir explícita y concretamente con los estudiantes los comportamientos que se interpretarán como trampa.  Los expertos alientan a los maestros a hacer que las tareas sean significativas para los estudiantes y a explicar el valor de lo que los estudiantes aprenderán al completarlas.

**Entorno de aprendizaje colaborativo**

 Un entorno de aprendizaje colaborativo ocurre cada vez que un instructor requiere que los estudiantes trabajen juntos en actividades de aprendizaje.  Los entornos de aprendizaje colaborativo pueden implicar actividades tanto formales como informales y pueden incluir o no una evaluación directa.  Por ejemplo, parejas de estudiantes trabajan en asignaciones de programación; pequeños grupos de estudiantes discuten posibles respuestas a la pregunta de un profesor durante la lección; y los estudiantes trabajan juntos fuera de clase para aprender nuevos conceptos.  El aprendizaje colaborativo es distinto de los proyectos en los que los estudiantes "dividen y vencerán".  Cuando los estudiantes dividen el trabajo, cada uno es responsable de solo una parte de la resolución del problema y hay oportunidades muy limitadas para resolver los problemas con los demás.  En entornos colaborativos, los estudiantes participan en conversaciones intelectuales entre ellos.



### <FONT COLOR="blue">**Q1.** </FONT>

El número natural e, tambien llamado el número de Euler o la constante de Napier, en un número irracional usado como la base de los logaritmos naturales y es definido con la siguiente formula:

$$e = 1 + \frac{1}{1!} + \frac{1}{2!} + \frac{1}{3!} + ... + \frac{1}{n!}$$

donde el valor de Euler depende de la cantidad de términos n que se establezcan. Desarrolle una función recursiva que permita encontrar el valor de euler(20) con 20 términos para encontrar los primeros 5 decimales.

*Salida esperada para n=20, es euler(20) = 2.71828*